// postcss-utils/lib/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = {
  ellipsis: function ellipsis(rows) {
    var rules = {
      'overflow': 'hidden',
      'text-overflow': 'ellipsis'
    };

    if (rows && rows > 1) {
      rules['display'] = '-webkit-box';
      rules['-webkit-box-orient'] = 'vertical';
      rules['-webkit-line-clamp'] = rows;
    } else {
      rules['white-space'] = 'nowrap';
    }

    return rules;
  },
  clearfix: function clearfix() {
    return {
      '&:before, &:after': {
        'display': 'table',
        'content': '""'
      },
      '&:after': {
        'clear': 'both'
      }
    };
  },
  irt: function irt() {
    return {
      'font': '0/0 none',
      'text-shadow': 'none',
      'color': 'transparent'
    };
  },
  userSelect: function userSelect(value) {
    return {
      '-moz-user-select': value,
      '-webkit-user-select': value,
      '-ms-user-select': value
    };
  },
  disabled: function disabled(bgc, bdc, color) {
    return {
      'background-color': bgc,
      'border-color': bdc,
      'color': color,
      'cursor': 'default',
      'pointer-events': 'none'
    };
  },
  verticalCenter: function verticalCenter() {
    return {
      '&:after': {
        'display': 'inline-block',
        'content': '""',
        'height': '100%',
        'vertical-align': 'middle'
      }
    };
  }
};
